### ISSUE TITLE
Issues With Multithreaded Applications

### ISSUE THREAD

**Issue by JonnyWhatshisface** [libevent/libevent#309]
Okay, I had closed it thinking it was my code. But at this point, I'm fairly certain it is, in fact, a libevent issue.

I've since added evbuffer_lock()'s prior to touching the evbuffer in any way, minus freeing, and I'm still getting the issue from within libevent.

Assertion ctx failed in evmap_io_active also appears from time to time,

```
(lldb) run
Process 45436 launched: '/Users/jonathanhall/Google Drive/libwebsock3/examples/echo' (x86_64)
libwebsock Version: 2.0.0 listening on port 8080 (sockfd: 3)
Process 45436 stopped
* thread #2: tid = 0x9c632, 0x0000000100070de5 libevent-2.0.5.dylib`bufferevent_free + 36 at bufferevent.c:763, stop reason = EXC_BAD_ACCESS (code=1, address=0x38)
    frame #0: 0x0000000100070de5 libevent-2.0.5.dylib`bufferevent_free + 36 at bufferevent.c:763
   760 		union bufferevent_ctrl_data d;
   761 		memset(&d, 0, sizeof(d));
   762 		BEV_LOCK(bev);
-> 763 		if (bev->be_ops->ctrl)
   764 			bev->be_ops->ctrl(bev, BEV_CTRL_CANCEL_ALL, &d);
   765 		BEV_UNLOCK(bev);
   766 	}
(lldb) bt
* thread #2: tid = 0x9c632, 0x0000000100070de5 libevent-2.0.5.dylib`bufferevent_free + 36 at bufferevent.c:763, stop reason = EXC_BAD_ACCESS (code=1, address=0x38)
  * frame #0: 0x0000000100070de5 libevent-2.0.5.dylib`bufferevent_free + 36 at bufferevent.c:763
    frame #1: 0x0000000100070dc1 libevent-2.0.5.dylib`bufferevent_free(bufev=0x00000001028000c0) + 49
    frame #2: 0x00000001000058bf libwebsock.2.dylib`lws_handle_close(job=0x000000010030d680) + 79 at websock.c:559
    frame #3: 0x0000000100004cf4 libwebsock.2.dylib`lws_worker_agent_function(ptr=0x0000000100300e90) + 212 at core.c:203
    frame #4: 0x00007fff91b9305a libsystem_pthread.dylib`_pthread_body + 131
    frame #5: 0x00007fff91b92fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #6: 0x00007fff91b903ed libsystem_pthread.dylib`thread_start + 13
(lldb) bt
* thread #2: tid = 0x9c632, 0x0000000100070de5 libevent-2.0.5.dylib`bufferevent_free + 36 at bufferevent.c:763, stop reason = EXC_BAD_ACCESS (code=1, address=0x38)
  * frame #0: 0x0000000100070de5 libevent-2.0.5.dylib`bufferevent_free + 36 at bufferevent.c:763
    frame #1: 0x0000000100070dc1 libevent-2.0.5.dylib`bufferevent_free(bufev=0x00000001028000c0) + 49
    frame #2: 0x00000001000058bf libwebsock.2.dylib`lws_handle_close(job=0x000000010030d680) + 79 at websock.c:559
    frame #3: 0x0000000100004cf4 libwebsock.2.dylib`lws_worker_agent_function(ptr=0x0000000100300e90) + 212 at core.c:203
    frame #4: 0x00007fff91b9305a libsystem_pthread.dylib`_pthread_body + 131
    frame #5: 0x00007fff91b92fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #6: 0x00007fff91b903ed libsystem_pthread.dylib`thread_start + 13
(lldb) 
```

I can either give you the code to libwebsock2 on a github if you'd like, or I can write up a reproducer for it. Let me know which you'd prefer.

**Comment by JonnyWhatshisface** [comment]
To give an idea of what I'm doing - I have four threads, each that has its own libevent base. I have four other threads that are running the event_loop's for those bases. The events trigger a function that adds a job to the scheduler, and the worker threads process the job then free the job. The problem is coming when I hit a buffer error / close. If I connect with roughly 130 connections and send garbage data (i.e. ctrl-Z) it causes the crash. If I connect and send normal data (i.e. a few new lines), the problem doesn't happen as frequently. 

As far as I can tell, it doesn't appear to ever even be reaching the ls_handle_close function prior to the crash.

**Comment by JonnyWhatshisface** [comment]
Little more to help...

```
(lldb) target create "examples/echo"
Current executable set to 'examples/echo' (x86_64).
(lldb) breakpoint set -n malloc_error_break 
Breakpoint 1: where = libsystem_malloc.dylib`malloc_error_break, address = 0x000000000001308e
(lldb) run
Process 47006 launched: '/Users/jonathanhall/Google Drive/libwebsock3/examples/echo' (x86_64)
libwebsock Version: 2.0.0 listening on port 8080 (sockfd: 3)
[err] evmap.c:401: Assertion ctx failed in evmap_io_active
Process 47006 stopped
* thread #2: tid = 0x9dc53, 0x00007fff9524b286 libsystem_kernel.dylib`__pthread_kill + 10, stop reason = signal SIGABRT
    frame #0: 0x00007fff9524b286 libsystem_kernel.dylib`__pthread_kill + 10
libsystem_kernel.dylib`__pthread_kill:
->  0x7fff9524b286 <+10>: jae    0x7fff9524b290            ; <+20>
    0x7fff9524b288 <+12>: movq   %rax, %rdi
    0x7fff9524b28b <+15>: jmp    0x7fff95246c53            ; cerror_nocancel
    0x7fff9524b290 <+20>: retq   
(lldb) bt
* thread #2: tid = 0x9dc53, 0x00007fff9524b286 libsystem_kernel.dylib`__pthread_kill + 10, stop reason = signal SIGABRT
  * frame #0: 0x00007fff9524b286 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff91b949f9 libsystem_pthread.dylib`pthread_kill + 90
    frame #2: 0x00007fff973709b3 libsystem_c.dylib`abort + 129
    frame #3: 0x0000000100076e5a libevent-2.0.5.dylib`event_exit(errcode=<unavailable>) + 42 at log.c:79
    frame #4: 0x00000001000771cb libevent-2.0.5.dylib`event_errx(eval=<unavailable>, fmt=<unavailable>) + 155 at log.c:136
    frame #5: 0x0000000100076609 libevent-2.0.5.dylib`evmap_io_active(base=<unavailable>, fd=<unavailable>, events=<unavailable>) + 185 at evmap.c:401
    frame #6: 0x000000010007abed libevent-2.0.5.dylib`kq_dispatch(base=0x0000000100300200, tv=<unavailable>) + 813 at kqueue.c:385
    frame #7: 0x00000001000666de libevent-2.0.5.dylib`event_base_loop(base=0x0000000100300200, flags=0) + 670 at event.c:1625
    frame #8: 0x00000001000041ef libwebsock.2.dylib`lws_evbase_thread_loop_function(ptr=0x00000001003001c0) + 15 at core.c:139
    frame #9: 0x00007fff91b9305a libsystem_pthread.dylib`_pthread_body + 131
    frame #10: 0x00007fff91b92fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #11: 0x00007fff91b903ed libsystem_pthread.dylib`thread_start + 13
(lldb) 
```

**Comment by JonnyWhatshisface** [comment]
Is it possible this is being caused because a thread that did not create the event base is running the event base loop? Should I have the thread the event base was created in run the loop?

**Comment by ellzey** [comment]
`evbuffer_lock` will not do anything unless you enable thread support; _before_ any other libevent functions are called. Have you done this?

> Is it possible this is being caused because a thread that did not create the event base is running the event base loop? Should I have the thread the event base was created in run the loop?

I don't really see why you wouldn't. That would be the most straight-forward and easier to manage.

**Comment by ellzey** [comment]
After looking over another issue you put in, again with the libwebsocket code; you seem to have a lot of thread-based issues that I've never encountered, or has been reported before. But I'm not playing the blame game: there may definitely be some issue. 

The other ticket had an argument value of 0x40, definitely some thread stack smashing going on. Would love to see some dtrace (I see you're using osx) and a -ggdb -O0 linked version of libevent.

Also what version?

Do you want me to look over your websocket code?

**Comment by JonnyWhatshisface** [comment]
azat - I'm now using 2.1.5-beta, which I'm experiencing the same thing. I'm making a few tweaks again trying to cover my end, and if I can't get it to behave in the next hour or two, I'll put the code up on github for you to take a look at. It would be and most certainly is appreciated; I'd be happy to hear I'm doing something silly and wonky versus there being an issue with libevent. :)

As a heads up, I'm enabling thread support at the very started of the code with evthread_use_pthreads(). Each buffer event is also started with BEV_OPT_THREADSAFE

**Comment by ellzey** [comment]
> azat - I'm now 

I am not @azat.

> As a heads up, I'm enabling thread support at the very started of the code with evthread_use_pthreads(). Each buffer event is also started with BEV_OPT_THREADSAFE 

Shouldn't need to do this if the evthread_use_pthreads has been set already.

My suggestion is to monitor your reference counts, maybe manually inc/dec if you plan on keeping events un freed for a duration.

**Comment by azat** [comment]
Shouldn't you?

https://github.com/libevent/libevent/blob/8d34302956001110643346447bbe926924e14d3f/bufferevent.c#L330
```
#ifndef EVENT__DISABLE_THREAD_SUPPORT
	if (options & BEV_OPT_THREADSAFE) {
		if (bufferevent_enable_locking_(bufev, NULL) < 0) {
                    ...
		}
	}
#endif
```

**Comment by ellzey** [comment]
```C
void
evbuffer_lock(struct evbuffer *buf)
{
        EVBUFFER_LOCK(buf);
}

#define EVBUFFER_LOCK(buffer)                                           \
        do {                                                            \
                EVLOCK_LOCK((buffer)->lock, 0);                         \
        } while (0)

#define EVLOCK_LOCK(lockvar,mode)                                       \
        do {                                                            \
                if (lockvar)                                            \
                        evthread_lock_fns_.lock(mode, lockvar);

**Comment by azat** [comment]
But `buffer->lock` is NULL there, and will be non-null only after `evbuffer_enable_locking()`

**Comment by JonnyWhatshisface** [comment]
ellzey - so sorry about that! It has been a long week for me. I'm a little brain dead.

I'm doing a little more digging, and I'm seeing something else that's kind of odd...
```
    input = bufferevent_get_input(bev);
    while ((n = evbuffer_remove(input, buf, sizeof(buf))) > 0) {
        printf("Data is: %s\n", buf);
    }
```
In my callback for the buffer event is the above. If I connect locally with telnet and type in abcdefg and hit enter:
```
Data is: abcdefg

```
If I press enter one more time without adding any new data:

```
Data is: abcd
g



```

Is this normal... ?

**Comment by JonnyWhatshisface** [comment]
Took the \n out of the printf() and I'm still getting inconsistent results:

```
Data is: abcdefg
Data is: 
cdefg
Data is: 
cdefg
Data is: 
cdefg
```

```
Data is: abcdefg
g eventing event loop!
Data is: 
cdefg
g eventing event loop!
Data is: 
cdefg
g eventing event loop!
```

**Comment by JonnyWhatshisface** [comment]
Please ignore the previous comments. I've been up more than 16 hours solid now... 

ellzey - thank you for the link. I'll read through it.

~J

**Comment by ellzey** [comment]
> In my callback for the buffer event is the above. If I con, each tienect locally with telnet and type in abcdefg and > hit enter:

> Data is: abcdefg

Try `printf("Data is: %,*s\n", (int)evbuffer_get_length(buf), buf);`

**Comment by ellzey** [comment]
@azat 

> But buffer->lock is NULL there, and will be non-null only after evbuffer_enable_locking()

Will there, without setting up `evthread_lock_fns_`? I'm confused.

```C
#define EVTHREAD_ALLOC_LOCK(lockvar, locktype)          \
        ((lockvar) = evthread_lock_fns_.alloc ?         \
            evthread_lock_fns_.alloc(locktype) : NULL)

/* ..... */

GLOBAL struct evthread_lock_callbacks evthread_lock_fns_ = {
        0, 0, NULL, NULL, NULL, NULL
};

```

**Comment by ellzey** [comment]
Also, full thread backtrace would be nice. I setup this quick alias in my gdbinit

```
define tbt
  thread apply all bt
end
document tbt
  Print backtrace of all running threads
end
```

just typ tbt

**Comment by azat** [comment]
@ellzey

> > But buffer->lock is NULL there, and will be non-null only after evbuffer_enable_locking()
> 
> Will there, without setting up `evthread_lock_fns_`? I'm confused.

I mean that buffer->lock will be non-NULL only after:
1) evthread_use_pthreads()
2) evbuffer_enable_locking()

IOW, just evthread_use_pthreads() is not enough to make buffer thread-safe.

**Comment by ellzey** [comment]
> IOW, just evthread_use_pthreads() is not enough to make buffer thread-safe.

True, but what I said was 

>> evbuffer_lock will not do anything unless you enable thread support; before any other libevent functions are called. Have you done this?

So we're on the same page.

**Comment by ellzey** [comment]
@JonnyWhatshisface FYI, if you get `libwebsocket2` working (_non-threaded_) and under a BSD compatible license, I would love to embed it in [Libevhtp](https://github.com/ellzey/libevhtp) (which will eventually work itself into Libevent, one day.)

**Comment by JonnyWhatshisface** [comment]
Most certainly... I've actually gotten it straightened out. As funny as it is - the issue was that I was free'ing the event buffer, and also closing the socket fd... I am using BEV_OPT_CLOSE_ON_FREE ... So that was generating the problems.

The threads are used solely to dispatch messages in a queue, as well as handle memory allocations etc for incoming connections. they're also configurable and can be adjusted to be single threaded. :)

So if you're interested, the message dispatching etc is now working. All I have left at this point is to add in the ability to send messages, put the ping for keep-alives in and write the send_message_by_tag function (I'm allowing a unique identifier to be set to each socket state so you can send by the unique id instead of having to reference the state, though you can stil send by the state just as well).

I should have something up by the end of the weekend. I'll have one or two memory leaks that I'll need to clean (I noted the ones I know of already) and then some bug testing.

Thank you guys for your assistance - that one little mishap had me ripping my hair out. The close() call was only in *one* place, and I managed to overlook it every time I checked the function.

**Comment by JonnyWhatshisface** [comment]
Also, I have no problem putting libwebsock under a BSD license.

**Comment by ellzey** [comment]
>  I have no problem putting libwebsock under a BSD license.

Sweet, is it state driven or ifelseifelse / whole blocks of data required to continue?

Also, you could just break out the base parser, and license that under BSD, any external type code (like things that call your parser) can be whatever. We would definitely import that. 

Hmm, _bufferevent_ws_? Mmmmm. :+1: 

Very interested.

**Comment by JonnyWhatshisface** [comment]
ellzey - It's state-driven. The bufcerevents process the data but put messages to be dispatched in to the scheduler for the worker threads to grab and dispatch them out. The beauty with it being state-driven is it can handle receiving and sending control frames in the middle of sending/receiving data frames, which the control frames have separate agents to process them. 

If you remember Payden Sutherlands libwebsock, it's modelled off of his, with some enhancements. I'm looking in to handing some more work to the dispatch threads with regards to handling the data parsing, as I ran a stress test today and bombarded messages through it. After cancelling the messages being sent in, I noticed libevent was still processing messages for a good 15 seconds after the fact. I'm thinking this is because I'm locking some of the structures with mutex's and it's holding the entire event thread up. So I'm going to tinker with handing the data parsing off to the dispatch threads to speed up the deliverance of the event callbacks.

It's still VERY rough draft right now, but if you'd like to take a look at it, i'll stick it up on github. I'm also going to have to do a lot of code clean up - it's a bit messy atm. Also, I have not added the code to free the frames on message dispatch yet - so she's eating memory like a fiend. 

As stated, a lot of it is built identical to Payden's libwebsock, just a slew of changes. 

If you're interested, I have something else I'd like to show you offline that you may take a liking to... It could be a very fun, and potentially profitable, project... HTML5/JS/PHP site I developed, which is going to rely on websockets in the long run, but uses XMLRPC for now as a means of developing the functional UI. I think it may interest you the final goal of that project... :)

libwebsock2 at its current state: https://github.com/JonnyWhatshisface/libwebsock2

**Comment by ellzey** [comment]
Ah ok. Seems like the parser is tied too close together with the main app. I started writing a zero copy FSM a while back, but I never stepped out of prototype 

https://github.com/ellzey/libevhtp/blob/libevhtp2/src/evhtp2/ws/evhtp_ws.c
